cmd/lotus-shed/fip-0036.go

```diff
@@ -0,0 +1,565 @@
+		// a map of client addresses to deal data stored in proposals
+		clientToDealStorage := make(map[address.Address]abi.StoragePower)
+
+		fmt.Println("iterating over all actors")
+		count := 0
+		err = st.ForEach(func(addr address.Address, act *types.Actor) error {
+			if count%200000 == 0 {
+				fmt.Println("processed ", count, " actors building maps")
+			}
+			count++
+			if builtin.IsMultisigActor(act.Code) {
+				ms, err := multisig.Load(store, act)
+				if err != nil {
+					return fmt.Errorf("load msig failed %v", err)
+
+				}
+
+				// TODO: Confirm that these are always ID addresses
+				signers, err := ms.Signers()
+				if err != nil {
+					return xerrors.Errorf("fail to get msig signers", err)
+				}
+				for _, s := range signers {
+					signerId := lookupId(s)
+					if m, found := accountsToMultisigs[signerId]; found { //add msig id to signer's collection
+						m = append(m, addr)
+						accountsToMultisigs[signerId] = m
+					} else {
+						n := []address.Address{addr}
+						accountsToMultisigs[signerId] = n
+					}
+				}
+
+				locked, err := ms.LockedBalance(abi.ChainEpoch(height))
+				if err != nil {
+					return xerrors.Errorf("failed to compute locked multisig balance: %w", err)
+				}
+
+				threshold, _ := ms.Threshold()
+				info := msigBriefInfo{
+					ID:        addr,
+					Signer:    signers,
+					Balance:   big.Min(big.Zero(), types.BigSub(act.Balance, locked)),
+					Threshold: threshold,
+				}
+				msigActorsInfo[addr] = info
+			}
+
+			if builtin.IsStorageMinerActor(act.Code) {
+				m, err := miner.Load(store, act)
+				if err != nil {
+					return xerrors.Errorf("fail to load miner actor: \n", err)
+				}
+
+				info, err := m.Info()
+				if err != nil {
+					return xerrors.Errorf("fail to get miner info: \n", err)
+				}
+
+				ownerId := lookupId(info.Owner)
+				if m, found := ownerMap[ownerId]; found { //add miner id to owner list
+					m = append(m, addr)
+					ownerMap[ownerId] = m
+				} else {
+					n := []address.Address{addr}
+					ownerMap[ownerId] = n
+				}
+
+				workerId := lookupId(info.Worker)
+				if m, found := workerMap[workerId]; found { //add miner id to worker list
+					m = append(m, addr)
+					workerMap[workerId] = m
+				} else {
+					n := []address.Address{addr}
+					workerMap[workerId] = n
+				}
+
+				bal, err := m.AvailableBalance(act.Balance)
+				if err != nil {
+					return err
+				}
+
+				pow, ok, err := powerState.MinerPower(addr)
+				if err != nil {
+					return err
+				}
+
+				if !ok {
+					pow.RawBytePower = big.Zero()
+				}
+
+				minerActorsInfo[addr] = minerBriefInfo{
+					rawBytePower: pow.RawBytePower,
+					// gets added up outside this loop
+					dealPower: big.Zero(),
+					balance:   bal,
+				}
+			}
+
+			return nil
+		})
+
+		if err != nil {
+			return err
+		}
+
+		fmt.Println("iterating over proposals")
+		dps, _ := marketState.Proposals()
+		if err := dps.ForEach(func(dealID abi.DealID, d market.DealProposal) error {
```

Comment:
@arajasek considering every deal as "legit" is not accurate. This is the filtering I do on my end:
`... d.sector_activation_epoch IS NOT NULL AND d.deal_slash_epoch IS NULL and d.end_epoch > 2162760`

It doesn't affect the result much at all, but still...

Reply:
Yeah, there's some nitpicking to do here. I'm not convinced that your approach is correcter (we should clarify all of this for future polls).

- A Deal in the `Proposals` already has is "accepted" by both parties and has collateral locked, so I'm not sure we should check `d.sector_activation_epoch IS NOT NULL`?
- A slashed deal is still a deal that a client made (and SP accepted) and both put up collateral for, so should it not be counted?
- `d.end_epoch > 2162760` excludes deals that have expired but cron hasn't GCed yet...that's probably a better definition of "active deals" but it's a narrow boundary?

I definitely don't think any of your suggestions are "wrong", just...not clearly correcter?

Reply:
(would be easy to adopt any subset of your suggestions, though)

Reply:
@arajasek answering in the same order

* The reason I always do this when parsing the state, is because we have deal Proposals that are slated to start far in the future, some in the next millennium
<details><summary>Expand for urgh</summary>

```
 deal_id  |  status   |    deal_start_time     | start_epoch 
----------+-----------+------------------------+-------------
   311683 | published | 3534-01-31 14:40:00+00 |  1591979600
   219097 | published | 3390-03-05 22:00:00+00 |  1440604800
   218449 | published | 3389-10-06 22:00:00+00 |  1440172800
   166347 | published | 3279-01-31 22:00:00+00 |  1323748800
   165999 | published | 3272-04-27 22:00:00+00 |  1316635200
   165922 | published | 3270-07-02 22:00:00+00 |  1314720000
   165919 | published | 3270-06-27 22:00:00+00 |  1314705600
   165917 | published | 3270-06-27 22:00:00+00 |  1314705600
   165921 | published | 3270-06-27 22:00:00+00 |  1314705600
   165918 | published | 3270-06-27 22:00:00+00 |  1314705600
   165912 | published | 3270-06-22 22:00:00+00 |  1314691200
   165916 | published | 3270-06-22 22:00:00+00 |  1314691200
   165911 | published | 3270-06-22 22:00:00+00 |  1314691200
   165915 | published | 3270-06-22 22:00:00+00 |  1314691200
   165914 | published | 3270-06-22 22:00:00+00 |  1314691200
   165908 | published | 3270-06-17 22:00:00+00 |  1314676800
   165909 | published | 3270-06-17 22:00:00+00 |  1314676800
   165905 | published | 3270-06-17 22:00:00+00 |  1314676800
   165910 | published | 3270-06-17 22:00:00+00 |  1314676800
   165907 | published | 3270-06-17 22:00:00+00 |  1314676800
   165904 | published | 3270-06-12 22:00:00+00 |  1314662400
   165903 | published | 3270-06-12 22:00:00+00 |  1314662400
   165902 | published | 3270-06-12 22:00:00+00 |  1314662400
   159205 | published | 3240-01-31 22:00:00+00 |  1282723200
   159203 | published | 3240-01-26 22:00:00+00 |  1282708800
   159200 | published | 3240-01-26 22:00:00+00 |  1282708800
   159204 | published | 3240-01-26 22:00:00+00 |  1282708800
   159202 | published | 3240-01-26 22:00:00+00 |  1282708800
   159201 | published | 3240-01-26 22:00:00+00 |  1282708800
   159197 | published | 3240-01-21 22:00:00+00 |  1282694400
   159195 | published | 3240-01-21 22:00:00+00 |  1282694400
   159196 | published | 3240-01-21 22:00:00+00 |  1282694400
   159193 | published | 3240-01-21 22:00:00+00 |  1282694400
   159194 | published | 3240-01-21 22:00:00+00 |  1282694400
   159190 | published | 3240-01-16 22:00:00+00 |  1282680000
   159191 | published | 3240-01-16 22:00:00+00 |  1282680000
   159192 | published | 3240-01-16 22:00:00+00 |  1282680000
   159188 | published | 3240-01-16 22:00:00+00 |  1282680000
   159189 | published | 3240-01-16 22:00:00+00 |  1282680000
   159183 | published | 3239-12-22 22:00:00+00 |  1282608000
   159181 | published | 3239-12-22 22:00:00+00 |  1282608000
   159170 | published | 3239-06-10 22:00:00+00 |  1282046400
   159155 | published | 3239-01-21 22:00:00+00 |  1281643200
   192728 | published | 2959-10-11 20:08:30+00 |   987868577
  2768016 | published | 2033-01-30 16:31:30+00 |    13080303
  2768015 | published | 2033-01-30 16:31:00+00 |    13080302
  2768014 | published | 2033-01-30 16:30:30+00 |    13080301
 11512152 | published | 2022-10-16 09:14:00+00 |     2253508
 11458042 | published | 2022-10-16 09:13:30+00 |     2253507
 11464093 | published | 2022-10-16 09:13:30+00 |     2253507
 11512158 | published | 2022-10-16 09:13:00+00 |     2253506
...
```
</details>

* "Slashed deal" is a misnomer: it means the underlying sector is gone for good and the deal has entered cron cleanup. I did confirm this with @ZenGround0 at multiple points.

* No strong feelings about this one, though note that cron cleanup is sometimes up to 24h

Reply:
> The reason I always do this when parsing the state, is because we have deal Proposals that are slated to start far in the future, some in the next millennium

Persuaded, adding that check.

> "Slashed deal" is a misnomer: it means the underlying sector is gone for good and the deal has entered cron cleanup. I did confirm this with @ZenGround0 at multiple points.

Yeah, you're right, but it's another 24 hour issue...I don't care enough for that ;)